### PR TITLE
Add promlinter to lint metrics name

### PR DESCRIPTION
**Pull request by yeya24**
Signed-off-by: Ben Ye <yb532204897@gmail.com>

This pr adds a linter for Prometheus metric naming. https://github.com/yeya24/promlinter

### CODE REVIEW

**Comment by sayboras** [comment]
@yeya24 awesome, let me know if you need any help :+1:

**Comment by sayboras** [comment]
Mark as draft PR to avoid showing up in review list.

**Comment by yeya24** [comment]
Sorry for the lag. Working on this pr now.

**Comment by yeya24** [comment]
@sayboras Sorry for the delay. This is ready for review now.

File: pkg/golinters/promlinter.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+			mu.Lock()
+			resIssues = append(resIssues, res...)
+			mu.Unlock()
+
+			return nil, nil
+		}
+	}).WithIssuesReporter(func(*linter.Context) []goanalysis.Issue {
+		return resIssues
+	}).WithLoadMode(goanalysis.LoadModeNone)
```
Comment (ldez) [inline pkg/golinters/promlinter.go:73]:
```suggestion
	}).WithLoadMode(goanalysis.LoadModeSyntax)
```

File: pkg/lint/lintersdb/manager.go [hunk 2]
```diff
@@ -479,7 +479,9 @@ func (m Manager) GetAllSupportedLinterConfigs() []*linter.Config {
 			WithSince("v1.39.0").
 			WithPresets(linter.PresetStyle, linter.PresetModule).
 			WithURL("https://github.com/ldez/gomoddirectives"),
-
+		linter.NewConfig(golinters.NewPromlinter()).
+			WithPresets(linter.PresetStyle).
+			WithURL("https://github.com/yeya24/promlinter"),
```
Comment (ldez) [inline pkg/lint/lintersdb/manager.go:484]:
```suggestion
		linter.NewConfig(golinters.NewPromlinter()).
			WithSince("v1.40.0").
			WithPresets(linter.PresetStyle).
			WithURL("https://github.com/yeya24/promlinter"),
```

**Review by ldez — CHANGES_REQUESTED** [review]

File: .golangci.example.yml [hunk 3]
```diff
@@ -347,6 +347,20 @@ linters-settings:
     simple: true
     range-loops: true # Report preallocation suggestions on range loops, true by default
     for-loops: false # Report preallocation suggestions on for loops, false by default
+  promlinter:
+    # Promlinter cannot infer all metrics name in static analysis. Enable strict mode
+    # will also include the errors caused by failing to parse the args.
+    strict: false
+    disabled-linters:
+    # Please refer to https://github.com/yeya24/promlinter#usage for detailed usage.
```
Comment (ldez) [inline .golangci.example.yml:355]:
```suggestion
    # Please refer to https://github.com/yeya24/promlinter#usage for detailed usage.
    disabled-linters:
```

File: .golangci.example.yml [hunk 4]
```diff
@@ -347,6 +347,20 @@ linters-settings:
     simple: true
     range-loops: true # Report preallocation suggestions on range loops, true by default
     for-loops: false # Report preallocation suggestions on for loops, false by default
+  promlinter:
+    # Promlinter cannot infer all metrics name in static analysis. Enable strict mode
+    # will also include the errors caused by failing to parse the args.
```
Comment (ldez) [inline .golangci.example.yml:352]:
```suggestion
    # Promlinter cannot infer all metrics name in static analysis.
    # Enable strict mode will also include the errors caused by failing to parse the args.
```

**Review by ldez — CHANGES_REQUESTED** [review]
I recommend using tags (ex: v0.1.0) instead of pseudo-version (v0.0.0-20210328235706-000c7d74ddb3) like that dependabot will be able to track the changes.

**Comment by yeya24** [comment]
@ldez Thanks for the quick and detailed review. I have updated my pr, PTAL.

**Review by bombsimon — APPROVED** [review]
Nice, LGTM

**Comment by yeya24** [comment]
Hi, is this pr good to go? Or anything else I should change?

File: pkg/golinters/promlinter.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+		nil,
+	).WithContextSetter(func(lintCtx *linter.Context) {
+		strict := lintCtx.Cfg.LintersSettings.Promlinter.Strict
+		disabledLinters := lintCtx.Cfg.LintersSettings.Promlinter.DisabledLinters
+
+		analyzer.Run = func(pass *analysis.Pass) (interface{}, error) {
+			files := make([]*ast.File, 0)
+
+			for _, f := range pass.Files {
+				if strings.HasSuffix(pass.Fset.Position(f.Pos()).Filename, "_test.go") {
```
Comment (ldez) [inline pkg/golinters/promlinter.go:39]: Could you explain why you have hardcoded a filter on test files?

**Review by ldez — CHANGES_REQUESTED** [review]

File: pkg/golinters/promlinter.go (hunk 5 above)
Comment (yeya24) [reply pkg/golinters/promlinter.go:39]: Oh, I want to skip the test files. Is there any other way to do this?

Comment (ldez) [reply pkg/golinters/promlinter.go:39]: yes there are several other ways:

```yml
# global approach
run:
  tests: false
```

or

```yml
# specific for a linter
issues:
  exclude-rules:
    - path: .*_test.go
      linters:
        - promlinter
```

or add an option to the linter.

Comment (yeya24) [reply pkg/golinters/promlinter.go]: Thanks, I have updated my code.

**Pull request merged**